### PR TITLE
machines: add sliders in Create VM dialog for memory and disk size

### PR DESCRIPTION
**Pull request by KKoukiou**
This allows us to remove the text bellow the input fields for
recommended and minumum values, since the slider incorporates these
information.

The tests had to be adjusted to ensured that the limits on the sliders
are respected.

### CODE REVIEW

**Comment by KKoukiou** [comment]
@garrett can you please help me with fixing the slider row?
I guess we need to keep the role=group and adjust to be applied to three elements. But I had trouble with that.

**Comment by KKoukiou** [comment]
@garrett it looks great - thanks a lot for the fixes. Can we do something better for mobile view?
The things is that the input text and the dropdown will take all the row size, and the rest will be allocated to the slider but it's already too little. Maybe we need some auto-fit/auto-fill and repeat?

**Comment by garrett** [comment]
Here are common phone sizes simulated in Firefox:

Pixel 2XL|iPhone X|iPhone 6|iPhone SE
---|---|---|---
slider-pixel2xl|slider-iphoneX|slider-iphone6|slider-iphoneSE

(A lot of Android phones also follow these various sizes, so it's more than just "iPhone". That said, not many phones in active use are as tiny as the SE these days.)

It only really becomes a problem for the SE, at least for visualization. It might be a bit fiddly to interact with on all of these otherwise.

Also, I think we'd need to drop the 3D effect from PF3 here too. I've been doing one-off hacks as I'm not sure where else we use the PF3 slider at the moment (if anywhere). Some will need to be one-offs, others perhaps should be moved into our PF3 overrides (if we use it elsewhere).

**Comment by garrett** [comment]
I reworked the sliders and made them a PF3 override for all PF3 sliders. The style is based on an older (but still the most recent) PF4 mockup:
54612563-8e21aa00-4a2f-11e9-8e03-cbbd89e225e4

Changes from the mockup:
- PF4 colors: the colors have been changed a few times in PF4 since that mockup was made.
- PF4 font sizes: I'm using the PF4 `xs` size for the text labels, which is closest to what's in the mockup.
- Sizing of widgets: I've based the sizes on rems (so it's tied to the font size) instead of pixels... this is especially important as the mockup is not pixel-perfect, so I had to approximate anyway.
- Focus/activation/hover: There's no activation state for a div, even when it has tabindex. There is focus and hover, so that's what I had to key things on.
- Focus ring: I intentionally used blank space between the ring and the thumb instead of using white. This allows the slider bar's color to peek through.

Anyway, the idea is that when PF4 sliders _finally_ arrive, they'll probably look and work similar to this. :wink:

**Comment by garrett** [comment]
Here's a screenshot that shows a slider in action, with its focus state and a tooltip. 

(I've hidden the background page with `visibility: hidden` in Firefox's dev tools so there's no extra noise. :wink:)

I've already optimized this with pngquant, so it should be ready for inclusion in the release announcement too.

vm-sliders

**Comment by marusak** [comment]
May I just ask an obvious questions?
Why `Size` and `Memory` lines do not take all the space as `Storage` or any other items do. I thought that the slider would just span.

**Comment by KKoukiou** [comment]
> May I just ask an obvious questions?
> Why `Size` and `Memory` lines do not take all the space as `Storage` or any other items do. I thought that the slider would just span.

@garrett redirecting to you

File: pkg/machines/components/memorySelectRow.jsx [hunk 1]
```diff
@@ -18,36 +18,78 @@
  */
 import React from 'react';
 import cockpit from 'cockpit';
+import { Slider } from 'patternfly-react';
```
[the remark this answers is not in the document]
Comment (garrett) [reply pkg/machines/components/memorySelectRow.jsx]: There is no PF4 slider. Yet.

Our options were to use the PF3 slider or the native `<input type="range">`

File: pkg/machines/components/memorySelectRow.css [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    transform: translateX(-50%);
+    /* Don't trap mouse hovering */
+    pointer-events: none;
+}
+
+/* Sadly, the PF3 slider is wrapped in an empty <div> */
+.slider-input-group > :first-child:not([class]) {
+    flex: auto;
+    min-width: 8rem;
+    max-width: 15rem;
```
[the remark this answers is not in the document]
Comment (garrett) [reply pkg/machines/components/memorySelectRow.css]: This is intended. If it's too wide, the input fields get lost visually and the slider becomes harder to use.

**Pull request merged**